src/assets/src/components/common.tsx

```diff
@@ -290,25 +245,93 @@ export const BackendSelector: React.FC<BackendSelectorProps> = (props) => {
     );
 }
 
-interface SingleInputFormShowRemainingProps extends StatelessSingleInputFormProps {
-    maxLength: number
+
+interface ValidatedInputFormProps extends SingleInputFormProps {
+    maxLength?: number;
+    validateLength?: boolean;
+    allowBlank: boolean;
 }
 
-export const SingleInputFormShowRemaining: React.FC<SingleInputFormShowRemainingProps> = (props) => {
-    const buttonClass = `btn btn-${props.buttonType} remaining-controls`;
-    const remaining = props.maxLength - props.value.length;
-    const isInvalid = props.value.length > props.maxLength;
+interface StatelessValidatedInputFormProps extends ValidatedInputFormProps {
+    value: string;
+    onChangeValue: (value: string) => void;
+}
+
+function generateFeedbackText (maxLength: number, currentLength: number, allowBlank: boolean) {
+    const remaining = maxLength - currentLength;
+    const charsRemaining = (remaining > 0) ? remaining : 0;
+    if (!allowBlank && currentLength === 0) {
+       return 'This field is not allowed to be blank.';
+    } else {
+        const charsOver = (remaining < 0) ? ` (${remaining * -1} over limit)` : '';
+        return `Remaining characters: ${charsRemaining}/${maxLength}${charsOver}`;
+    }
+}
+
+export const StatelessInputGroupForm: React.FC<StatelessValidatedInputFormProps> = (props) => {
+    const buttonClass = `btn btn-${props.buttonType}`
+
+    let feedbackEl;
+    let isInvalid;
+    if (props.validateLength && props.maxLength !== undefined) {
+        isInvalid = props.value.length > props.maxLength || (!props.allowBlank && props.value.length === 0);
+        const feedbackText = generateFeedbackText(props.maxLength, props.value.length, props.allowBlank);
+        const textClass = isInvalid ? 'text-danger': '';
+        feedbackEl = (
+            <Form.Text bsPrefix={`form-text remaining-feedback ${textClass}`}>
+                {feedbackText}
+            </Form.Text>
+        );
+    }
```

Comment:
Feels like there's a little room for improvement here... Maybe validation logic could be factored out of here or made more modular somehow (Passing in a validate function? Passing in some kind of validationOptions object?).

Reply:
You are correct, I need to refactor the validation work here. I'm looking into using [`yup`](https://github.com/jquense/yup#string) for validation (which was used in a React Bootstrap example), which I think might be a nice way to make this more customizable. I'll update once I've had more time to experiment.

Reply:
Okay, I think I like how this has progressed. I moved validation logic into a new `validation.ts` file, and I've adopted `yup` to create schemas with error messages. I want to do some more testing on whether multiple feedback messages are possible, but it seems potentially useful to keep that option open even if `schema.validate` only ever returns one message.